File: services/python-service/controllers/admin_aparcament_controller.py
"""
Controlador d'operacions administratives per a la gestió d'aparcaments (CRUD + imatges).

Protegit per autenticació JWT amb rol d'administrador o operador. Gestiona el
cicle de vida complet dels aparcaments: llistat paginator, creació, actualització
i eliminació. El processament d'imatges es delega al model `admin_aparcament_model`.
"""

from flask import jsonify, request
from models.admin_aparcament_model import (
    get_admin_aparcaments,
    count_admin_aparcaments,
    create_aparcament,
    update_aparcament,
    delete_aparcament,
    save_parking_images,
)
from middleware.jwt_auth import get_jwt_user_id


def get_admin_list():
    """
    GET /api/admin/aparcaments - Llista els aparcaments amb filtres i paginació.

    Query params:
        search (str): Cerca per nom, adreça o ciutat.
        type (str): Filtre per tipologia.
        status (str): Filtre per estat (actiu/inactiu).
        page (int): Número de pàgina (per defecte 1).
        limit (int): Registres per pàgina (màx. 100, per defecte 10).

    Returns:
        JSON 200: Llista d'aparcaments i informació de paginació.
        JSON 500: Error intern del servidor.
    """
    try:
        search = request.args.get('search', '')
        tipo = request.args.get('type', '')
        status = request.args.get('status', '')
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 10))

        if page < 1:
            page = 1
        if limit < 1 or limit > 100:
            limit = 10

        offset = (page - 1) * limit

        total = count_admin_aparcaments(search, tipo, status)
        data = get_admin_aparcaments(search, tipo, status, limit, offset)

        total_pages = (total + limit - 1) // limit

        return jsonify({
            'success': True,
            'data': data,
            'pagination': {
                'total': total,
                'page': page,
                'limit': limit,
                'total_pages': total_pages
            }
        }), 200

    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


def create_admin():
    """
    POST /api/admin/aparcaments - Crea un nou aparcament (multipart/form-data).

    Camps obligatoris del formulari: nom, tipus, adreca, ciutat, latitud, longitud.
    Camps opcionals: imatges via 'parking_images[]', tarifes, horaris, etc.
    Si el processament d'imatges falla, el rollback elimina l'aparcament creat.

    Returns:
        JSON 201: ID de l'aparcament creat.
        JSON 400: Si falten camps obligatoris o les imatges són invàlides.
        JSON 500: Error intern del servidor.
    """
    try:
        # Obtenir dades del formulari (multipart/form-data)
        data = request.form.to_dict()

        # Validar dades obligatòries
        required = ['nom', 'tipus', 'adreca', 'ciutat', 'latitud', 'longitud']
        missing = [field for field in required if not data.get(field)]
        if missing:
            return jsonify({
                'success': False,
                'error': f"Falten dades obligatòries: {', '.join(missing)}"
            }), 400

        # Crear aparcament
        parking_id = create_aparcament(data)

        # Procesar imatges si n'hi ha
        user_id = None
        try:
            user_id = get_jwt_user_id()
        except:
            pass

        if 'parking_images[]' in request.files:
            try:
                save_parking_images(
                    parking_id,
                    request.files,
                    user_id
                )
            except ValueError as ve:
                # Si falla el upload de fotos, eliminar l'aparcament creat
                delete_aparcament(parking_id)
                return jsonify({'success': False, 'error': str(ve)}), 400

        return jsonify({
            'success': True,
            'message': 'Aparcament creat correctament',
            'id': parking_id
        }), 201

    except ValueError as ve:
        return jsonify({'success': False, 'error': str(ve)}), 400
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


def update_admin():
    """
    PUT /api/admin/aparcaments - Actualitza les dades d'un aparcament existent.

    Query params:
        id (int): ID de l'aparcament a modificar (obligatori).

    Camps del formulari: mateixos que create_admin.
    Si s'envien imatges noves via 'parking_images[]', s'afegeixen a les existents.

    Returns:
        JSON 200: Confirmació de l'actualització.
        JSON 400: Si l'ID falta, no és vàlid o falten camps obligatoris.
        JSON 500: Error intern del servidor.
    """
    try:
        parking_id = request.args.get('id')
        if not parking_id:
            return jsonify({'success': False, 'error': 'ID d\'aparcament no proporcionat'}), 400

        try:
            parking_id = int(parking_id)
        except ValueError:
            return jsonify({'success': False, 'error': 'ID d\'aparcament no vàlid'}), 400

        # Obtenir dades del formulari
        data = request.form.to_dict()

        # Validar dades obligatòries
        required = ['nom', 'tipus', 'adreca', 'ciutat', 'latitud', 'longitud']
        missing = [field for field in required if not data.get(field)]
        if missing:
            return jsonify({
                'success': False,
                'error': f"Falten dades obligatòries: {', '.join(missing)}"
            }), 400

        # Actualitzar aparcament
        update_aparcament(parking_id, data)

        # Procesar imatges si n'hi ha
        user_id = None
        try:
            user_id = get_jwt_user_id()
        except:
            pass

        if 'parking_images[]' in request.files:
            try:
                save_parking_images(
                    parking_id,
                    request.files,
                    user_id
                )
            except ValueError as ve:
                return jsonify({'success': False, 'error': str(ve)}), 400

        return jsonify({
            'success': True,
            'message': 'Aparcament actualitzat correctament'
        }), 200

    except ValueError as ve:
        return jsonify({'success': False, 'error': str(ve)}), 400
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


def delete_admin():
    """
    DELETE /api/admin/aparcaments - Elimina un aparcament i tots els seus recursos.

    Query params:
        id (int): ID de l'aparcament a eliminar (obligatori).

    Returns:
        JSON 200: Confirmació de l'eliminació.
        JSON 400: Si l'ID falta o no és vàlid.
        JSON 500: Error intern del servidor.
    """
    try:
        parking_id = request.args.get('id')
        if not parking_id:
            return jsonify({'success': False, 'error': 'ID d\'aparcament no proporcionat'}), 400

        try:
            parking_id = int(parking_id)
        except ValueError:
            return jsonify({'success': False, 'error': 'ID d\'aparcament no vàlid'}), 400

        delete_aparcament(parking_id)

        return jsonify({
            'success': True,
            'message': 'Aparcament eliminat correctament'
        }), 200

    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

File: services/php-service/controllers/AdminAparcamentController.php
<?php

require_once __DIR__ . "/../models/AdminAparcamentModel.php";
require_once __DIR__ . '/../middleware/AuthMiddleware.php';

class AdminAparcamentController
{
    private $model;

    public function __construct()
    {
    }

    private static function isAdmin()
    {
        $usuari = AuthMiddleware::obtenirUsuariAutenticat();
        if (!$usuari || !isset($usuari['tipus_usuari'])) return false;

        $rol = strtolower($usuari['tipus_usuari']);
        return $rol === 'administrador' || $rol === 'admin';
    }

    public function processRequest()
    {
        AuthMiddleware::verificarAutenticacio();
        if (!self::isAdmin()) {
            $this->respond(['success' => false, 'error' => 'No tens permisos per realitzar aquesta acció'], 403);
        }

        $this->model = new AdminAparcamentModel();

        $method = $_SERVER['REQUEST_METHOD'];
        $action = $_GET['action'] ?? '';

        switch ($method) {
            case 'GET':
                $search = trim($_GET['search'] ?? '');
                $type = trim($_GET['type'] ?? '');
                $status = trim($_GET['status'] ?? '');

                $page = isset($_GET['page']) ? max(1, (int)$_GET['page']) : 1;
                $limit = isset($_GET['limit']) ? max(1, (int)$_GET['limit']) : 10;
                $offset = ($page - 1) * $limit;

                $total = $this->model->getTotalAparcamentsCount($search, $type, $status);
                $parkings = $this->model->getAllAparcaments($search, $type, $status, $limit, $offset);

                if (!empty($this->model->getErrors())) {
                    $this->respond([
                        'success' => false,
                        'error' => 'Error de dades',
                        'details' => $this->model->getErrors()
                    ], 500);
                }

                $totalPages = ceil($total / $limit);

                $this->respond([
                    'success' => true,
                    'data' => $parkings,
                    'pagination' => [
                        'total' => $total,
                        'page' => $page,
                        'limit' => $limit,
                        'total_pages' => $totalPages
                    ]
                ]);
                break;

            case 'POST':
                $contentType = $_SERVER['CONTENT_TYPE'] ?? '';
                $isMultipart = stripos($contentType, 'multipart/form-data') !== false;

                if ($isMultipart) {
                    $data = $_POST;
                } else {
                    $data = json_decode(file_get_contents('php://input'), true);
                }

                if (!is_array($data)) {
                    $data = [];
                }

                $uploadedImages = $_FILES['parking_images'] ?? null;
                $userId = AuthMiddleware::obtenirIdUsuari();

                if ($action === 'create') {
                    $this->handleCreate($data, $uploadedImages, $userId);
                } elseif ($action === 'update') {
                    $id = $_GET['id'] ?? null;
                    $this->handleUpdate($id, $data, $uploadedImages, $userId);
                } elseif ($action === 'delete') {
                    $id = $_GET['id'] ?? null;
                    $this->handleDelete($id);
                } else {
                    $this->respond(['success' => false, 'error' => 'Acció no vàlida'], 400);
                }
                break;

            default:
                $this->respond(['success' => false, 'error' => 'Mètode no permès'], 405);
                break;
        }
    }

    private function handleCreate($data, $uploadedImages = null, $userId = null)
    {
        if (empty($data['nom']) || empty($data['tipus']) || empty($data['adreca']) || empty($data['latitud']) || empty($data['longitud'])) {
            $this->respond(['success' => false, 'error' => 'Falten dades obligatòries'], 400);
        }

        $result = $this->model->createAparcament($data, $uploadedImages, $userId);
        if ($result) {
            $this->respond(['success' => true, 'message' => 'Aparcament creat correctament', 'id' => $result]);
        } else {
            $this->respond(['success' => false, 'errors' => $this->model->getErrors()], 500);
        }
    }

    private function handleUpdate($id, $data, $uploadedImages = null, $userId = null)
    {
        if (!$id) {
            $this->respond(['success' => false, 'error' => 'ID d\'aparcament no proporcionat'], 400);
        }

        $result = $this->model->updateAparcament($id, $data, $uploadedImages, $userId);
        if ($result) {
            $this->respond(['success' => true, 'message' => 'Aparcament actualitzat correctament']);
        } else {
            $this->respond(['success' => false, 'errors' => $this->model->getErrors()], 500);
        }
    }

    private function handleDelete($id)
    {
        if (!$id) {
            $this->respond(['success' => false, 'error' => 'ID d\'aparcament no proporcionat'], 400);
        }

        $result = $this->model->deleteAparcament($id);
        if ($result) {
            $this->respond(['success' => true, 'message' => 'Aparcament eliminat correctament']);
        } else {
            $this->respond(['success' => false, 'errors' => $this->model->getErrors()], 500);
        }
    }

    private function respond($data, $status = 200)
    {
        http_response_code($status);
        header('Content-Type: application/json');
        echo json_encode($data);
        exit();
    }
}

if (basename($_SERVER['PHP_SELF']) === basename(__FILE__)) {
    new AdminAparcamentController();
}
